### DISCUSSION TITLE
ConversableAgent "send" function "context" question

### DISCUSSION THREAD

**Discussion by bonadio** [microsoft/autogen#458] [Q&A] [answered]
Hi 
I was looking at the "send" function in ConversableAgent,

[conversable_agent](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent/)

But I did not understand the example with the lambda function

`{
    "content": lambda context: context["use_tool_msg"],
    "context": {
        "use_tool_msg": "Use tool X if they are relevant."
    }
}`

If I run this code 

`user_proxy.send( { 
            "content": lambda context: context["use_tool_msg"],
            "context": {
                "use_tool_msg": "Use tool X if they are relevant."
            }}, order_assistant, request_reply=True)`

I get 

`user_proxy (to order_assistant):

Use tool X if they are relevant.

 `

What is the use case for this type of message?

Thanks

**Comment by sonichi** [comment]
@JieyuZ2 would you like to explain this use case?

**Comment by JieyuZ2** [comment, marked as answer]
The example is to show how to use the context mechanism. In general, the context mechanism allows you to change any user message in the conversation history. For example, the context could be


```
def content_function(use_b):
    if use_b:
        return str_a + str_b
    else:
        return str_a  
```

Then, the first message you send to the model could be 

```
{ "content": content_function, "context": { "use_b": True } }
```

At certain turn of the conversation, you want to switch the `first message` to `str_a` only, you can send 


```
{ "content": "current message", "context": { "use_b": False } }
```

Then, when calling the LLM, the first message would be changed to `str_a` only.   (note: when using chat LLMs, every time we have to send it the whole conversation history.)

One use case of this mechanism is, the initial message contains some useful but long information (`str_b`), and you only want to send this information to the LLM once. After the LLM's first reply, you can turn off the `str_b` by sending a new context with your current message to the LLM to save the cost and token used.

In my EcoAssistant project (https://github.com/JieyuZ2/EcoAssistant/), I append some demonstrations to the first message sent to the LLM, then in the subsequent conversation / calls of LLM, I turn off the demonstration in the first message to save the number of inputting tokens.

Let me know if you have any question!

**Reply by bonadio** [reply]
Hi @JieyuZ2, your note that every time we send the whole conversation, was what I was missing. 
The code in your project is really good. You should post at the Discord channel in the Tutorials. 
Thanks

**Comment by zuoyifan132** [comment]
@JieyuZ2 I wonder if send and receive function can be used in an instance of autogen.GroupChat? Since the chat is managed by the group chat manager, is there a way to intervene the group chat?

**Reply by JieyuZ2** [reply]
If by intervening the group chat you mean selecting next speaker manually, then you can set the speaker selection to `manual` (https://github.com/microsoft/autogen/blob/main/autogen/agentchat/groupchat.py#L41)

If you mean provide human feedback, you can use a userproxy agent in groupchat, and send human feedback via the userproxy agent just like in basic two-agent chat

**Reply by zuoyifan132** [reply]
I want to create a small group chat within a larger group chat. However, I don't want the messages from the small group to be visible in the larger group's chat history. Only the final results from the small group chat should be shared with the larger group. I'm trying to find a way to do this using the send function.

**Reply by JieyuZ2** [reply]
I think this new society_of_mind agent may fit your need
https://github.com/microsoft/autogen/blob/main/notebook/agentchat_society_of_mind.ipynb

**Reply by zuoyifan132** [reply]
Thanks for the solution and that perfectly solve my issue. One concern about the autogen framework: does the framework deal with the long context(long group chat history since all agents share with same history) problem?

**Reply by JieyuZ2** [reply]
this may help https://github.com/microsoft/autogen/blob/main/notebook/agentchat_compression.ipynb

**Reply by sonichi** [reply]
@gagb has made a new way to handle long context for any agent using a new capability.
cc @kevin666aa

**Reply by zuoyifan132** [reply]
@JieyuZ2 does the outside world chat history all be visible from SocietyOfMind internal monologue group?
